lib/DxilValidation/DxilValidation.cpp

```diff
@@ -970,6 +970,226 @@ static void ValidateImmOperandForMathDxilOp(CallInst *CI, DXIL::OpCode Opcode,
+    ValCtx.EmitInstrError(
+        CI, ValidationRule::InstrLinalgInvalidRegisterInterpValue);
+    return;
+  }
+
+  auto MIValue = MI->getLimitedValue();
+  if (!CheckInMemoryInterpretations(MIValue)) {
+    ValCtx.EmitInstrError(CI,
+                          ValidationRule::InstrLinalgInvalidMemoryInterpValue);
+    return;
+  }
+
+  ConstantInt *MatrixTransposeConst = dyn_cast<ConstantInt>(MatrixTranspose);
+  ConstantInt *MatrixLayoutConst = dyn_cast<ConstantInt>(MatrixLayout);
+  if (!llvm::isa<llvm::Constant>(MatrixM) ||
+      !llvm::isa<llvm::Constant>(MatrixK) || !MatrixLayoutConst ||
+      !MatrixTransposeConst) {
+    ValCtx.EmitInstrError(CI,
+                          ValidationRule::InstrLinalgMatrixShapeParamsAreConst);
+    return;
+  }
+
+  auto MLValue = MatrixLayoutConst->getLimitedValue();
+  if (!CheckMatrixLayout(MLValue)) {
+    ValCtx.EmitInstrError(CI,
+                          ValidationRule::InstrLinalgInvalidMatrixLayoutValue);
+    return;
+  }
+
+  if (!CheckTransposeForMatrixLayout(
+          static_cast<DXIL::DXILMatrixLayout>(MLValue),
+          MatrixTransposeConst->getLimitedValue())) {
+    ValCtx.EmitInstrError(
+        CI, ValidationRule::InstrLinalgMatrixLayoutNotTransposable);
+    return;
+  }
+
+  llvm::Value *InputVector =
+      CI->getOperand(DXIL::OperandIndex::kMatVecMulInputVectorIdx);
+  if (!CheckUnsignedFlag(InputVector->getType(),
+                         IsInputUnsignedConst->getLimitedValue())) {
+    ValCtx.EmitInstrError(CI, ValidationRule::InstrLinalgNotAnUnsignedType);
+    return;
+  }
+
+  if (!CheckUnsignedFlag(CI->getType(),
+                         IsOutputUnsignedConst->getLimitedValue())) {
+    ValCtx.EmitInstrError(CI, ValidationRule::InstrLinalgNotAnUnsignedType);
+    return;
+  }
+
+  switch (OpCode) {
+  case DXIL::OpCode::MatVecMulAdd: {
+    llvm::Value *BiasInterpretation =
+        CI->getOperand(DXIL::OperandIndex::kMatVecMulAddBiasInterpretation);
+    if (!llvm::isa<llvm::Constant>(BiasInterpretation)) {
+      ValCtx.EmitInstrError(
+          CI, ValidationRule::InstrLinalgInterpretationParamAreConst);
+      return;
+    }
+    ConstantInt *BI = cast<ConstantInt>(BiasInterpretation);
+    auto BIValue = BI->getLimitedValue();
+    if (!CheckInMemoryInterpretations(BIValue)) {
+      ValCtx.EmitInstrError(
+          CI, ValidationRule::InstrLinalgInvalidMemoryInterpValue);
+      return;
+    }
+  } break;
+  default:
+    break;
+  }
+}
+
+static void ValidateImmOperandsForOuterProdAcc(CallInst *CI,
+                                               ValidationContext &ValCtx) {
+
+  llvm::Value *MatrixInterpretation =
+      CI->getOperand(DXIL::OperandIndex::kOuterProdAccMatrixInterpretation);
+  llvm::Value *MatrixLayout =
+      CI->getOperand(DXIL::OperandIndex::kOuterProdAccMatrixLayout);
+
+  if (!llvm::isa<llvm::Constant>(MatrixInterpretation)) {
+    ValCtx.EmitInstrError(
+        CI, ValidationRule::InstrLinalgInterpretationParamAreConst);
+  }
```

Comment:
nit: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements
```suggestion
  if (!llvm::isa<llvm::Constant>(MatrixInterpretation))
    ValCtx.EmitInstrError(
        CI, ValidationRule::InstrLinalgInterpretationParamAreConst);
```